oidc_middleware.go

```diff
@@ -0,0 +1,87 @@
+package main
+
+import (
+	"crypto/rsa"
+	"encoding/base64"
+	"encoding/json"
+	"github.com/gin-gonic/gin"
+	"math/big"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/golang-jwt/jwt"
+)
+// Largely taken from https://developer.okta.com/blog/2021/01/04/offline-jwt-validation-with-go
+
+func GetOidcPublicKeys(config OICDConfig) *map[string]*rsa.PublicKey {
+	rsakeys := make(map[string]*rsa.PublicKey)
+	var body map[string]interface{}
+	resp, err := http.Get(config.KeysURL)
+	if err != nil {
+		log.Error().WithError(err).Message("Could not fetch from KeysURL.")
+	}
+	err = json.NewDecoder(resp.Body).Decode(&body)
+	if err != nil {
+		log.Error().WithError(err).Message("Failed to decode login JWT Keys.")
+	}
+	log.Debug().Message("Updating keys for oidc.")
+	for _, bodykey := range body["keys"].([]interface{}) {
+		key := bodykey.(map[string]interface{})
+		kid := key["kid"].(string)
+		rsakey := new(rsa.PublicKey)
+		number, _ := base64.RawURLEncoding.DecodeString(key["n"].(string))
+		rsakey.N = new(big.Int).SetBytes(number)
+		rsakey.E = 65537
+		rsakeys[kid] = rsakey
+	}
+	return &rsakeys
+}
+
+func VerifyTokenMiddleware(config OICDConfig, rsakeys *map[string]*rsa.PublicKey) gin.HandlerFunc {
+	return func (ginContext *gin.Context) {
+		if *rsakeys == nil {
+			log.Warn().Message("RsaKeys for OIDC have not been set (http:500).")
+			ginContext.AbortWithStatus(http.StatusInternalServerError)
+		}
```

Comment:
Suggest validating this field during initialization instead of during runtime. Such as returning an error at the beginning of `VerifyTokenMiddleware`, alternatively toss this into a type and have it validate on initialization and then send a method into the `gin.Use`

Reply:
I think its either here or not at all. This works fine without the null check as well. Moving it up makes an an on init check only- which seems pointless to me, better to nullcheck it where relevant.

Reply:
Ah ok I missed that you update the map in the Ticker thing.

Strongly advice to toss this into a type then, so the map and function belongs together better.

Ex:

```go
type OIDCMiddleware struct {
	rsaKeys map[string]*rsa.PublicKey
}

func (m *OIDCMiddleware) HandlerFunc(c *gin.Context) {
	// verify request
}

func (m *OIDCMiddleware) AutoRefreshKeys() {
	ticker := time.NewTicker(time.Hour * 25)
	// refresh m.rsaKeys logic
}


	oidcMiddleware := &OIDCMiddleware{}
	r := gin.New()
	r.Use(oidcMiddleware.HandlerFunc)
```